### PR TITLE
Dev refactor config

### PR DESCRIPTION
**Pull request by alexbpx**

### CODE REVIEW

File: perimeterx/px_constants.py [hunk 1]
```diff
@@ -24,5 +25,7 @@
 EMPTY_GIF_B64 = 'R0lGODlhAQABAPAAAAAAAAAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw=='
 COLLECTOR_HOST = 'collector.perimeterx.net'
 FIRST_PARTY_FORWARDED_FOR = 'X-FORWARDED-FOR'
-
 MODULE_VERSION = 'Python WSGI Module v2.0.0'
+API_RISK = '/api/v2/risk'
```
Comment (yaronpx) [inline perimeterx/px_constants.py]: use v3

File: perimeterx/px_config.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            app_id.lower())
+        self._encryption_enabled = config_dict.get('encryption_enabled', True)
+        self._sensitive_headers = config_dict.get('sensitive_headers', ['cookie', 'cookies'])
+        self._send_page_activities = config_dict.get('send_page_activities', True)
+        self._api_timeout = config_dict.get('api_timeout', 500)
+        self._custom_logo = config_dict.get('custom_logo', '')
+        self._css_ref = config_dict.get('_custom_logo', '')
+        self._js_ref = config_dict.get('js_ref', '')
+        self._is_mobile = config_dict.get('is_mobile', False)
+        self._monitor_mode = 0 if module_mode is px_constants.MODULE_MODE_MONITORING else 1
```
[the remark this answers is not in the document]
Comment (alexbpx) [reply perimeterx/px_config.py]: Fixed

File: perimeterx/px_context.py [hunk 3]
```diff
@@ -67,14 +60,25 @@ def build_context(environ, config):
...
         'vid': vid,
-        'query_params': environ['QUERY_STRING']
+        'query_params': environ['QUERY_STRING'],
+        'sensitive_route': sensitive_route,
+        'whitelist_route': whitelist_route,
     }
     return ctx
 
+
 def extract_ip(config, environ):
```
Comment (ilaipx) [inline perimeterx/px_context.py]: move to a util module

File: perimeterx/px_cookie.py [hunk 4]
```diff
@@ -13,32 +13,32 @@
...
+            return PxCookieV1(ctx, self._config)
 
         if prefix == PREFIX_PX_COOKIE_V3:
-            config["logger"].debug("PxCookie[build_px_cookie] using cookie v3")
+            self._logger.debug("PxCookie[build_px_cookie] using cookie v3")
             from px_cookie_v3 import PxCookieV3
-            return PxCookieV3(ctx, config)
+            return PxCookieV3(ctx, self._config)
 
     def decode_cookie(self):
```
Comment (ilaipx) [inline perimeterx/px_cookie.py]: move to util

File: perimeterx/px_cookie.py [hunk 5]
```diff
@@ -13,32 +13,32 @@
...
-        self.config['logger'].debug("PxCookie[decode_cookie]")
+        self._logger.debug("PxCookie[decode_cookie]")
         return base64.b64decode(self.raw_cookie)
 
     def pbkdf2_hmac(self, hash_name, password, salt, iterations, dklen=None):
```
Comment (ilaipx) [inline perimeterx/px_cookie.py]: move to util

File: perimeterx/px_cookie.py [hunk 6]
```diff
@@ -142,28 +142,28 @@ def is_cookie_valid(self, str_to_hmac):
...
         try:
-            calculated_digest = hmac.new(self.config['cookie_key'], str_to_hmac, hashlib.sha256).hexdigest()
+            calculated_digest = hmac.new(self._config.cookie_key, str_to_hmac, hashlib.sha256).hexdigest()
             return self.get_hmac() == calculated_digest
         except:
-            self.config["logger"].debug("failed to calculate hmac")
+            self._logger.debug("failed to calculate hmac")
             return False
 
     def deserialize(self):
```
Comment (ilaipx) [inline perimeterx/px_cookie.py]: I would move every "generic" action to a util, and only leave methods that are specific to pxCookie (as per our discussion with the data enrichment that also shares some of the logic such as encode, decode, serialize, deserialize, encrypt, decrypt, etc).

File: perimeterx/px_context.py (hunk 3 above)
Comment (alexbpx) [reply perimeterx/px_context.py]: Not in this PR

File: perimeterx/middleware.py [hunk 7]
```diff
@@ -1,83 +1,70 @@
...
-            'first_party': True,
-            'first_party_xhr_enabled': True,
-            'monitor_mode': px_constants.MODULE_MODE_MONITORING,
-            }
-
-        self.config = dict(self.config.items() + config.items())
-        self.config['logger'] = logger = Logger(self.config['debug_mode'])
-        if not config['app_id']:
+        px_config = PXConfig(config)
+        logger = px_config.logger
```
[the remark this answers is not in the document]
Comment (alexbpx) [reply perimeterx/middleware.py]: Not in this PR

File: perimeterx/px_cookie.py (hunk 6 above)
Comment (alexbpx) [reply perimeterx/px_cookie.py]: That's a different PR

**Pull request merged**